v-next/hardhat/src/internal/builtin-plugins/network-manager/json-rpc-request-modifiers/chain-id/chain-id.ts

```diff
@@ -0,0 +1,50 @@
+import type { EthereumProvider } from "../../../../../types/providers.js";
+
+import { assertHardhatInvariant } from "@ignored/hardhat-vnext-errors";
+
+import { rpcQuantityToNumber } from "../utils.js";
+
+export abstract class ChainId {
+  protected readonly provider: EthereumProvider;
+
+  #chainId: number | undefined;
+
+  constructor(provider: EthereumProvider) {
+    this.provider = provider;
+  }
+
+  protected async getChainId(): Promise<number> {
+    if (this.#chainId === undefined) {
+      try {
+        this.#chainId = await this.getChainIdFromEthChainId();
+      } catch {
+        // If eth_chainId fails we default to net_version
+        this.#chainId = await this.#getChainIdFromEthNetVersion();
+      }
+    }
+
+    return this.#chainId;
+  }
+
+  protected async getChainIdFromEthChainId(): Promise<number> {
+    const id = await this.provider.request({
+      method: "eth_chainId",
+    });
+
+    assertHardhatInvariant(typeof id === "string", "id should be a string");
+
+    return rpcQuantityToNumber(id);
```

Comment:
In hh-utils we have the method `hexStringToBigInt` but not the method `hexStringToNumber`. Do we want to add the method `hexStringToNumber` in hh-utils?

Reply:
Done in [this commit](https://github.com/NomicFoundation/hardhat/pull/5810/commits/b7b80c1cd8ac73afa5ad5dc3e7b60bb037cd59c9)